### PR TITLE
enable the role to install geoipupdate and run updater every wednesday at 5am

### PR DESCRIPTION
**Pull request by amanet2**
This update will pull/install geoipupdate from apt, as well as copy over the config file we need for upgrading our specific geoip db.

High level view
===
This will change our ansible role for `elastic-stack-5.x` to tell a provisioned logstash machine to update its own geoip db every wednesday at 5am.

5c-roles will pull in this repo as a submodule `elastic-stack-5.x`

An easy way to run this is to clone `5c_playbooks` and `5c-roles` in the same working dir.  Clone this repo, and replace all the files in `5c-roles/elastic-stack-5.x` with this repo's files.  Go to `5c_playbooks/vagrant/logstash` and run `vagrant up`.

Alternatively, you can also create your own playbook and vagrantfile for running this repo.  Just make sure these files are in the same directory as `ansible-role-elastic-stack-5.x`

Istruzioni
===
## Playbook and Vagrantfile

playbook.yml

```yaml
- hosts: all
  roles:
    - role: ansible-role-elastic-stack-5.x
      install_java: yes
      install_logstash: yes
```

vagrantfile

```ruby
# -*- mode: ruby -*-
# vi: set ft=ruby :

Vagrant.configure('2') do |config|
  config.vm.box = 'ubuntu/trusty64'
  config.vm.hostname = 'elastic-stack'
  config.vm.network 'private_network', type: :dhcp

  config.vm.provider :virtualbox do |v, override|
    v.memory = 2048
    v.customize ['modifyvm', :id, '--natdnshostresolver1', 'on']
    v.customize ['modifyvm', :id, '--natdnsproxy1', 'on']
    v.customize ['modifyvm', :id, '--name', 'elastic-stack']
  end

  config.vm.provision :ansible do |ansible|
    ansible.playbook = 'playbook.yml'
  end
end
```

## Running geoipupdate from command line

1. After provisioning completes `vagrant ssh`
1. `sudo /usr/bin/geoipupdate -v -f /etc/GeoIP.conf`
1. The first one should download and unzip and archive
1. `sudo /usr/bin/geoipupdate -v -f /etc/GeoIP.conf`
1. The second one should give a msg along the lines of `Nothing to update`

## Checking for crontab file
1. `/var/spool/cron/crontabs/root` should exist and look like this: 
```
0 5 * * 3 /usr/bin/geoipupdate -v -f /etc/GeoIP.conf
```

Automation
===
Uses a cron tab to currently run every wednesday at 5.  The frequency of updates can be configured in `defaults/main.yml`

Note
===
If you do NOT see anything in the ansible output regarding geoipupdate, something failed and you should comment and get my attention.

### CODE REVIEW

File: tasks/logstash.yml [hunk 1]
```diff
@@ -119,6 +119,27 @@
   become_user: logstash
   when: ls_install_geoip and geolitecity_dat_gz.stat.exists
 
+- name: Download GeoIP Update
+  apt_repository: repo='ppa:maxmind/ppa'
+  become: yes
```
Comment (ajrgrubbs) [inline tasks/logstash.yml]: Don't use this syntax. Use the new style, which is easier to read:
```yaml
- name: Download GeoIP Update
  apt_repository:
    repo: ppa:maxmind/ppa
  become: yes
```

Also this step has an inaccurate name. You're adding the maxmind repository, not downloading an update. Should be something like `Add Maxmind PPA`

File: tasks/logstash.yml [hunk 2]
```diff
@@ -119,6 +119,27 @@
...
+
+- name: Update apt
+  apt: upgrade=yes
+  become: yes
+  become_user: root
```
Comment (ajrgrubbs) [inline tasks/logstash.yml]: This step is upgrading all the packages currently installed on the machine - which doesn't include the geoipupdate package. That's out of the scope of this role and should not happen.

Unless you meant to have it update its cache, which is handled with an extra parameter. These are easily confused, to be fair:
- Apt Update: Check all known repositories, index a list of packages which are available to install
- Apt Upgrade: Check for new versions of locally-installed software, upgrade them if necessary.

You probably want to delete this step.

File: tasks/logstash.yml [hunk 3]
```diff
@@ -119,6 +119,27 @@
...
+
+- name: install geoipupdate
+  apt: name='geoipupdate'
+  become: yes
```
Comment (ajrgrubbs) [inline tasks/logstash.yml]: Use the new style yaml, and tell it to update the cache

```yaml
- name: install geoipupdate
  apt:
    name: geoipupdate
    update_cache: yes
  become: yes
```

That step will install it if it is missing. If you want to ensure geoipupdate is also upgraded, you give it an extra argument: `state: latest`

File: tasks/logstash.yml [hunk 4]
```diff
@@ -119,6 +119,27 @@
...
+
+- name: copy geoipupdate config
+  copy:
+    src: "{{ item.src }}"
+    dest: "{{ item.dest }}"
+  with_items:
+    - {src: "{{ role_path }}/files/GeoIP.conf", dest: "{{ ls_yml['path.geoipconfig'] }}"}
+  become: yes
```
Comment (ajrgrubbs) [inline tasks/logstash.yml]: Using a `with` loop is overkill here - you only have one item.
```yaml
- name: copy geoipupdate config
  copy:
    src: "{{ role_path }}/files/GeoIP.conf"
    dest: "{{ ls_yml['path.geoipconfig'] }}"
  become: yes
```

File: tasks/logstash.yml [hunk 5]
```diff
@@ -157,6 +178,19 @@
   become: yes
   become_user: logstash
 
+- name: remove old cron
+  cron:
+    name: update_geoip
+    state: absent
```
Comment (ajrgrubbs) [inline tasks/logstash.yml]: If your experience tells differently, then ok - but a step like this shouldn't be necessary. According to [the documentation](http://docs.ansible.com/ansible/cron_module.html), since you are specifying `name: update_geoip` it should update the cron job if it already exists. Removing it first doesn't do us much.

File: tasks/logstash.yml [hunk 6]
```diff
@@ -157,6 +174,19 @@
   become: yes
   become_user: logstash
 
+# - name: remove old cron
+#   cron:
+#     name: update_geoip
+#     state: absent
```
Comment (ajrgrubbs) [inline tasks/logstash.yml]: Why comment this out instead of deleting it?

File: defaults/main.yml [hunk 7]
```diff
@@ -63,11 +63,16 @@ ls_yml:
...
+  path.geoipconfig: /etc/GeoIP.conf
   config.string: ""
   config.test_and_exit: "false"
   config.reload.automatic: "false"
   config.reload.interval: 3
   config.debug: "false"
+  config.geoipupdate_month: "*"
+  config.geoipupdate_weekday: "*"
+  config.geoipupdate_hour: "*"
+  config.geoipupdate_minute: "*"
```
Comment (ajrgrubbs) [inline defaults/main.yml]: Let's set a default of running every Wednesday at 5am.

**~Slightly~ Very convoluted justification**: According to Maxmind's docs, their databases are updated on Tuesdays (on either a weekly or monthly basis). They do not specify a time of day. They're based in Massachusetts, which is eastern time. So to ensure we run this after their database has been updated during the day, we would run it at 12am ET wednesday morning, or 1pm CT.
BUT we don't actually know if our servers will be in CT or UTC (the default). UTC is ahead of ET by either 4 or 5 hours, depending on whether we are in daylight savings time. So we assume the worst and say 5 hours - aka 5AM.

*In summary*: Fuck time zones, man

File: tasks/logstash.yml [hunk 8]
```diff
@@ -157,6 +174,19 @@
...
+
+- name: add cron job for geoipupdate
+  cron:
+    name: update_geoip
+    weekday: "{{ ls_yml['config.geoipupdate_weekday'] }}"
+    hour: "{{ ls_yml['config.geoipupdate_hour'] }}"
+    minute: "{{ ls_yml['config.geoipupdate_minute'] }}"
+    job: "sudo /usr/bin/geoipupdate -v -f {{ ls_yml['path.geoipconfig'] }}"
```
Comment (ajrgrubbs) [inline tasks/logstash.yml]: Remove `sudo` from the job, and make this step `become: yes` instead. Instead of running the job as a user which escalates it's privileges to root, let's just run it as root to begin with.

File: .gitignore [hunk 9]
```diff
@@ -1 +1,2 @@
 .idea
+*.DS_Store
```
Comment (ajrgrubbs) [inline .gitignore]: This kind of entry is perfect for a global gitignore file, fyi

File: tasks/main.yml [hunk 10]
```diff
@@ -12,6 +12,9 @@
 - include: logstash.yml
   when: install_logstash
 
+- include: geoip-update.yml
```
Comment (ajrgrubbs) [inline tasks/main.yml]: Nice refactor

**Comment by amanet2** [comment]
Actually, it seems to be updating the db in the wrong folder.  it's currently doing `/usr/share/geoip/...` instead of `etc/logstash/geodb/...`

I'll get this fixed tomorrow, very good catch

**Comment by ajrgrubbs** [comment]
It may actually be best if we remove the logic for downloading/unpacking the database, and instead pointing logstash at the file in its default `/usr/share/...` location. Think about what would be cleaner.

**Comment by amanet2** [comment]
I verified the cron file was there, and THIS time, there are no updates available on the first update attempt....which leads me to believe we're pointing to the right database this time.

File: files/GeoIP.conf [hunk 11]
```diff
@@ -0,0 +1,18 @@
+# The following UserId and LicenseKey are required placeholders:
+UserId 999999
+LicenseKey 000000000000
+
+# The remaining settings are OPTIONAL.
+
+# The directory to store the database files. Defaults to /usr/local/share/GeoIP
+DatabaseDirectory /etc/logstash/geodb
```
Comment (ajrgrubbs) [inline files/GeoIP.conf]: Let's get this parameter templatized too. You'll need to change the ansible task from "copy" to "template".

Comment (ajrgrubbs) [inline files/GeoIP.conf]: Oh - and since it's moving to a template it should move from `files/` to `templates/`
